Add Functionality to use RestoreData using Kopia Server (#1914)
kanisterio/kanister: docs/functions.rst

It creates a new Pod that mounts the PVCs referenced by the specified Pod
and restores data to the specified path.

.. note::
   It is extremely important that, the PVCs are not be currently
   in use by an active application container, as they are required
   to be mounted to the new Pod (ensure by using
   ``ScaleWorkload`` with replicas=0 first).
   For advanced use cases, it is possible to have concurrent access but
   the PV needs to have ``RWX`` mode enabled and the volume needs to use a
   clustered file system that supports concurrent access.

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

   `namespace`, Yes, `string`, namespace of the application that you want to restore the data in
   `backupIdentifier`, Yes, `string`, unique snapshot id generated during backup
   `restorePath`, Yes, `string`, path where data to be restored
   `pod`, No, `string`, pod to which the volumes are attached
   `volumes`, No, `map[string]string`, Mapping of `pvcName` to `mountPath` under which the volume will be available
   `podOverride`, No, `map[string]interface{}`, specs to override default pod specs with

.. note::

>    Between the ``pod`` and ``volumes`` arguments, exactly one argument
>    must be specified.

Reviewer: ```suggestion
   Either ``pod`` or the ``volumes`` arguments must be specified to this function based on the function that was used to backup the data. For `BackupData` we should specify `pod` and for `CopyVolumeData`, `Volumes` should be specified.
```

Feel free to change words/grammar according to you.

Author: LGTM

        args:
          name: "test-snapshot-content-content-dfc8fa67-8b11-4fdf-bf94-928589c2eed8"

.. _restoredatausingkopiaserver:

RestoreDataUsingKopiaServer
---------------------------

This function restores data backed up by the ``BackupDataUsingKopiaServer`` function.
It creates a new Pod that mounts the PVCs referenced by the Pod specified in the
function argument and restores data to the specified path.

.. note::
   It is extremely important that, the PVCs are not currently
   in use by an active application container, as they are required
   to be mounted to the new Pod (ensure by using
   ``ScaleWorkload`` with replicas=0 first).
   For advanced use cases, it is possible to have concurrent access but

>    the PV needs to have ``RWX`` access mode  and the volume needs to use a

Reviewer: ```suggestion
   the PV needs to have ``RWX`` access mode and the volume needs to use a
```

   clustered file system that supports concurrent access.

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

   `namespace`, Yes, `string`, namespace of the application that you want to restore the data in
   `image`, Yes, `string`, image to be used for running restore job (should contain kopia binary)
   `backupIdentifier`, Yes, `string`, unique snapshot id generated during backup
   `restorePath`, Yes, `string`, path where data to be restored
   `pod`, No, `string`, pod to which the volumes are attached

>    `volumes`, No, `map[string]string`, Mapping of `pvcName` to `mountPath` under which the volume will be available

Reviewer: ```suggestion
   `volumes`, No, `map[string]string`, mapping of `pvcName` to `mountPath` under which the volume will be available
```

   `podOverride`, No, `map[string]interface{}`, specs to override default pod specs with

.. note::
   The ``image`` argument requires the use of ``ghcr.io/kanisterio/kanister-tools``
   image since it includes the required tools to restore data from
   the object store.

   Either ``pod`` or the ``volumes`` arguments must be specified to this function
   based on the function that was used to backup the data.

>    Is `BackupDataUsingKopiaServer` is used to backup the data we should specify `pod` and for
>    `CopyVolumeDataUsingKopiaServer`, `Volumes` should be specified.

Reviewer 2: ```suggestion
   If `BackupDataUsingKopiaServer` is used to backup the data we should specify `pod` and for
   `CopyVolumeDataUsingKopiaServer`, `Volumes` should be specified.
```

   the PV needs to have ``RWX`` mode enabled and the volume needs to use a
   clustered file system that supports concurrent access.

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

   `namespace`, Yes, `string`, namespace of the application that you want to restore the data in
   `image`, Yes, `string`, image to be used for running restore job (should contain kopia binary)
   `backupIdentifier`, Yes, `string`, unique snapshot id generated during backup
   `restorePath`, Yes, `string`, path where data to be restored
   `userPassphrase`, Yes, `string`, user access credentials for kopia repository server
   `certData`, Yes, `string`, certificate data for kopia repository server
   `pod`, No, `string`, pod to which the volumes are attached
   `volumes`, No, `map[string]string`, Mapping of `pvcName` to `mountPath` under which the volume will be available
   `podOverride`, No, `map[string]interface{}`, specs to override default pod specs with

.. note::
   The ``image`` argument requires the use of ``ghcr.io/kanisterio/kanister-tools``
   image since it includes the required tools to restore data from
   the object store.

>    Between the ``pod`` and ``volumes`` arguments, exactly one argument

Author: The RestoreData in general is being used to restore data for BackupData and CopyVolumeData. 
So in case of Restoring snapsot created by BackupData we mention pod and
In case of of restoring snapshot created by CopyVolumeData we mention volumes

>    Between the ``pod`` and ``volumes`` arguments, exactly one argument

Author: You could check it [here](http://localhost:63342/kanister/docs/_build/html/functions.html?_ijt=3156deoqoo25u4o6rk44nvfjr6&_ij_reload=RELOAD_ON_SAVE#restoredata)

   must be specified.

>    Additionally, in order to use this function, a RepositoryServer CR is required

Reviewer: ```suggestion
   Additionally, in order to use this function, a RepositoryServer CR is required.
```

        args:
          name: "test-snapshot-content-content-dfc8fa67-8b11-4fdf-bf94-928589c2eed8"

.. _restoredatausingkopiaserver:

RestoreDataUsingKopiaServer
---------------------------

> This function restores data backed up by the BackupDataUsingKopiaServer function.

Reviewer: ```suggestion
This function restores data backed up by the ``BackupDataUsingKopiaServer`` function.
```

It creates a new Pod that mounts the PVCs referenced by the specified Pod
and restores data to the specified path.

.. note::
   It is extremely important that, the PVCs are not be currently
   in use by an active application container, as they are required
   to be mounted to the new Pod (ensure by using

>    ScaleWorkload with replicas=0 first).
>    For advanced use cases, it is possible to have concurrent access but
>    the PV needs to have RWX mode enabled and the volume needs to use a
>    clustered file system that supports concurrent access.

Reviewer: ```suggestion
   ``ScaleWorkload`` with replicas=0 first).
   For advanced use cases, it is possible to have concurrent access but
   the PV needs to have ``RWX`` mode enabled and the volume needs to use a
   clustered file system that supports concurrent access.
```

.. csv-table::
   :header: "Argument", "Required", "Type", "Description"
   :align: left
   :widths: 5,5,5,15

>    `namespace`, Yes, `string`, namespace in which to execute

Reviewer: same comments here, let's make sure users actually understand what these arguments are.
